roachtest: sqlsmith nil context

[(roachtest).sqlsmith/setup=rand-tables/setting=no-mutations failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2751365&tab=buildLog) on [master@15a185606d5e80b47d9fdd0ed4f54cfe29c527c6](https://github.com/cockroachdb/cockroach/commits/15a185606d5e80b47d9fdd0ed4f54cfe29c527c6):

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/sqlsmith/setup=rand-tables/setting=no-mutations/run_1
	cluster.go:1667,context.go:140,cluster.go:1656,test_runner.go:848: dead node detection: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod monitor teamcity-2751365-1615015616-07-n4cpu4 --oneshot --ignore-empty-nodes: exit status 1 1: 7036
		3: 6447
		4: dead
		2: dead
		Error: UNCLASSIFIED_PROBLEM: 4: dead
		(1) UNCLASSIFIED_PROBLEM
		Wraps: (2) secondary error attachment
		  | 2: dead
		  | (1) attached stack trace
		  |   -- stack trace:
		  |   | main.glob..func14
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1147
		  |   | main.wrap.func1
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:271
		  |   | github.com/spf13/cobra.(*Command).execute
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:830
		  |   | github.com/spf13/cobra.(*Command).ExecuteC
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
		  |   | github.com/spf13/cobra.(*Command).Execute
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
		  |   | main.main
		  |   | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1852
		  |   | runtime.main
		  |   | 	/usr/local/go/src/runtime/proc.go:204
		  |   | runtime.goexit
		  |   | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		  | Wraps: (2) 2: dead
		  | Error types: (1) *withstack.withStack (2) *errutil.leafError
		Wraps: (3) attached stack trace
		  -- stack trace:
		  | main.glob..func14
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1147
		  | main.wrap.func1
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:271
		  | github.com/spf13/cobra.(*Command).execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:830
		  | github.com/spf13/cobra.(*Command).ExecuteC
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:914
		  | github.com/spf13/cobra.(*Command).Execute
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/vendor/github.com/spf13/cobra/command.go:864
		  | main.main
		  | 	/home/agent/work/.go/src/github.com/cockroachdb/cockroach/pkg/cmd/roachprod/main.go:1852
		  | runtime.main
		  | 	/usr/local/go/src/runtime/proc.go:204
		  | runtime.goexit
		  | 	/usr/local/go/src/runtime/asm_amd64.s:1374
		Wraps: (4) 4: dead
		Error types: (1) errors.Unclassified (2) *secondary.withSecondaryError (3) *withstack.withStack (4) *errutil.leafError

```

<details><summary>More</summary><p>


Artifacts: [/sqlsmith/setup=rand-tables/setting=no-mutations](https://teamcity.cockroachdb.com/viewLog.html?buildId=2751365&tab=artifacts#/sqlsmith/setup=rand-tables/setting=no-mutations)

[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2Asqlsmith%2Fsetup%3Drand-tables%2Fsetting%3Dno-mutations.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

Looks like a dup of #58366 which has been fixed on this SHA, cc @asubiotto 
```
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x30 pc=0xe5e002]

goroutine 53300 [running]:
panic(0x44807a0, 0x7b57110)
	/usr/local/go/src/runtime/panic.go:1064 +0x545 fp=0xc001625310 sp=0xc001625248 pc=0x48b285
runtime.panicmem(...)
	/usr/local/go/src/runtime/panic.go:212
runtime.sigpanic()
	/usr/local/go/src/runtime/signal_unix.go:742 +0x413 fp=0xc001625340 sp=0xc001625310 pc=0x4a1fd3
github.com/cockroachdb/cockroach/pkg/util/tracing.SpanFromContext(0x0, 0x0, 0x7fc96b1b8d28)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/tracing/context.go:25 +0x22 fp=0xc001625378 sp=0xc001625340 pc=0xe5e002
github.com/cockroachdb/cockroach/pkg/util/log.getSpanOrEventLog(0x0, 0x0, 0x3, 0x462000, 0x546fce)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:94 +0x39 fp=0xc0016253b0 sp=0xc001625378 pc=0xf793f9
github.com/cockroachdb/cockroach/pkg/util/log.vEventf(0x0, 0x0, 0x0, 0x1, 0x2, 0x4c8416b, 0x57, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:242 +0x12a fp=0xc0016254e0 sp=0xc0016253b0 pc=0xf7a7ea
github.com/cockroachdb/cockroach/pkg/util/log.VEvent(...)
	/go/src/github.com/cockroachdb/cockroach/pkg/util/log/trace.go:263
github.com/cockroachdb/cockroach/pkg/sql/execinfra.MisplannedRanges(0x0, 0x0, 0xc0024ad590, 0x1, 0x1, 0x7fc900000004, 0xc000e73d40, 0x4306320, 0x40de858, 0x45e11f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/readerbase.go:66 +0x8c fp=0xc0016257b8 sp=0xc0016254e0 pc=0x235eeec
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateMeta(0xc002c82900, 0x0, 0x0, 0x5, 0x4b2a5e8, 0x4b2a5e8)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:314 +0x4ba fp=0xc001625990 sp=0xc0016257b8 pc=0x2600c5a
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateTrailingMeta(0xc002c82900, 0x5, 0xc002b3a501, 0x58d4760)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:175 +0x47 fp=0xc0016259e8 sp=0xc001625990 pc=0x25ffa67
github.com/cockroachdb/cockroach/pkg/sql/rowexec.(*tableReader).generateTrailingMeta-fm(0xc0016e0ab0, 0x1000000007486d6, 0x450c560)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/rowexec/tablereader.go:172 +0x2a fp=0xc001625a18 sp=0xc0016259e8 pc=0x26169aa
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).moveToTrailingMeta(0xc002c82900)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:727 +0x24e fp=0xc001625b58 sp=0xc001625a18 pc=0x235d64e
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).MoveToDraining(0xc002c82900, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:611 +0x1d8 fp=0xc001625c48 sp=0xc001625b58 pc=0x235cf58
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).ConsumerDone(0xc002c82900)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:942 +0x33 fp=0xc001625c70 sp=0xc001625c48 pc=0x235eb53
github.com/cockroachdb/cockroach/pkg/sql/execinfra.(*ProcessorBase).MoveToDraining(0xc002ce7b00, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfra/processorsbase.go:608 +0x1ad fp=0xc001625d60 sp=0xc001625c70 pc=0x235cf2d
github.com/cockroachdb/cockroach/pkg/sql/colexec.(*Columnarizer).DrainMeta(0xc002ce7b00, 0x596c200, 0xc001266450, 0x0, 0x0, 0x1)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colexec/columnarizer.go:252 +0x57 fp=0xc001625e18 sp=0xc001625d60 pc=0x2c1ff77
github.com/cockroachdb/cockroach/pkg/sql/execinfrapb.MetadataSources.DrainMeta(0xc000737690, 0x1, 0x1, 0x596c200, 0xc001266450, 0x85e84f8, 0xc00051ef28, 0x76ddf0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/execinfrapb/data.go:343 +0xbd fp=0xc001625ea8 sp=0xc001625e18 pc=0x224ea7d
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*HashRouter).Run(0xc0011c8790, 0x596c200, 0xc001266450)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/routers.go:614 +0xd1 fp=0xc001625f38 sp=0xc001625ea8 pc=0x3272131
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreator).setupRouter.func1(0x596c140, 0xc003311cc0, 0xc0007371c0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:730 +0xdb fp=0xc001625f90 sp=0xc001625f38 pc=0x327b9bb
github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1.1(0xc002c25f20, 0x596c140, 0xc003311cc0, 0xc0007371c0, 0xc002ef2560)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1279 +0x44 fp=0xc001625fb8 sp=0xc001625f90 pc=0x327e264
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1374 +0x1 fp=0xc001625fc0 sp=0xc001625fb8 pc=0x4c46c1
created by github.com/cockroachdb/cockroach/pkg/sql/colflow.(*vectorizedFlowCreatorHelper).accumulateAsyncComponent.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/colflow/vectorized_flow.go:1278 +0x6f
```
Possibly, #61347 is to blame if the table reader is never `Start`ed before being drained.

--- a/pkg/sql/colexec/colbuilder/execplan.go
+++ b/pkg/sql/colexec/colbuilder/execplan.go
@@ -61,8 +61,7 @@ func wrapRowSources(
 	flowCtx *execinfra.FlowCtx,
 	inputs []colexecop.Operator,
 	inputTypes [][]*types.T,
-	acc *mon.BoundAccount,
-	processorID int32,
+	args *colexecargs.NewColOperatorArgs,
 	newToWrap func([]execinfra.RowSource) (execinfra.RowSource, error),
 	factory coldata.ColumnFactory,
 ) (*colexec.Columnarizer, error) {
@@ -78,6 +77,17 @@ func wrapRowSources(
 			c.MarkAsRemovedFromFlow()
 			toWrapInputs = append(toWrapInputs, c.Input())
 		} else {
+			var metadataSources execinfrapb.MetadataSources
+			if len(args.MetadataSources) > i {
+				// In some testing paths, MetadataSources might be left unset,
+				// so we check whether the slice has ith element. In the
+				// production setting though the length of Inputs is always the
+				// same as the length of MetadataSources.
+				metadataSources = args.MetadataSources[i]
+				// We pass on the responsibility of draining metadata sources to
+				// the materializer.
+				args.MetadataSources[i] = nil
+			}
 			// Note that this materializer is *not* added to the set of
 			// releasables because in some cases it could be released before
 			// being closed. Namely, this would occur if we have a subquery
@@ -88,11 +98,11 @@ func wrapRowSources(
 			// when the main planNode tree is being closed.
 			toWrapInput, err := colexec.NewMaterializer(
 				flowCtx,
-				processorID,
+				args.Spec.ProcessorID,
 				input,
 				inputTypes[i],
 				nil, /* output */
-				nil, /* metadataSourcesQueue */
+				metadataSources,
 				nil, /* toClose */
 				nil, /* getStats */
 				nil, /* cancelFlow */
@@ -111,11 +121,11 @@ func wrapRowSources(
 
 	if _, mustBeStreaming := toWrap.(execinfra.StreamingProcessor); mustBeStreaming {
 		return colexec.NewStreamingColumnarizer(
-			ctx, colmem.NewAllocator(ctx, acc, factory), flowCtx, processorID, toWrap,
+			ctx, colmem.NewAllocator(ctx, args.StreamingMemAccount, factory), flowCtx, args.Spec.ProcessorID, toWrap,
 		)
 	}
 	return colexec.NewBufferingColumnarizer(
-		ctx, colmem.NewAllocator(ctx, acc, factory), flowCtx, processorID, toWrap,
+		ctx, colmem.NewAllocator(ctx, args.StreamingMemAccount, factory), flowCtx, args.Spec.ProcessorID, toWrap,
 	)
 }
 
@@ -561,8 +571,7 @@ func (r opResult) createAndWrapRowSource(
 		flowCtx,
 		inputs,
 		inputTypes,
-		args.StreamingMemAccount,
-		spec.ProcessorID,
+		args,
 		func(inputs []execinfra.RowSource) (execinfra.RowSource, error) {
 			// We provide a slice with a single nil as 'outputs' parameter
 			// because all processors expect a single output. Passing nil is ok
@@ -1344,6 +1353,12 @@ func NewColOperator(
 	r.Op, r.ColumnTypes = addProjection(r.Op, r.ColumnTypes, projection)
 	if args.TestingKnobs.PlanInvariantsCheckers {
 		r.Op = colexec.NewInvariantsChecker(r.Op)
+	}
+	// Handle the metadata sources from the input trees. Note that it is
+	// possible that we have created a materializer which took over draining
+	// the sources given to us by the caller.
+	for i := range args.MetadataSources {
+		r.MetadataSources = append(r.MetadataSources, args.MetadataSources[i]...)
 	}
 	return r, err
 }

--- a/pkg/sql/colexec/colexecargs/op_creation.go
+++ b/pkg/sql/colexec/colexecargs/op_creation.go
@@ -38,11 +38,17 @@ type NewColOperatorArgs struct {
 	StreamingMemAccount  *mon.BoundAccount
 	ProcessorConstructor execinfra.ProcessorConstructor
 	LocalProcessors      []execinfra.LocalProcessor
-	DiskQueueCfg         colcontainer.DiskQueueCfg
-	FDSemaphore          semaphore.Semaphore
-	ExprHelper           *ExprHelper
-	Factory              coldata.ColumnFactory
-	TestingKnobs         struct {
+	// MetadataSources are all sources of the metadata that are present in the
+	// trees rooted in Inputs. The slice has the same length as Inputs. If
+	// NewColOperator call creates a colexec.Materializer, then it will take
+	// over the responsibility of draining the sources; otherwise, they will be
+	// returned in NewColOperatorResult.
+	MetadataSources []execinfrapb.MetadataSources
+	DiskQueueCfg    colcontainer.DiskQueueCfg
+	FDSemaphore     semaphore.Semaphore
+	ExprHelper      *ExprHelper
+	Factory         coldata.ColumnFactory
+	TestingKnobs    struct {
 		// SpillingCallbackFn will be called when the spilling from an in-memory
 		// to disk-backed operator occurs. It should only be set in tests.
 		SpillingCallbackFn func()
@@ -82,10 +88,16 @@ type NewColOperatorArgs struct {
 // NewColOperatorResult is a helper struct that encompasses all of the return
 // values of NewColOperator call.
 type NewColOperatorResult struct {
-	Op              colexecop.Operator
-	KVReader        colexecop.KVReader
-	ColumnTypes     []*types.T
-	MetadataSources []execinfrapb.MetadataSource
+	Op          colexecop.Operator
+	KVReader    colexecop.KVReader
+	ColumnTypes []*types.T
+	// MetadataSources are all sources of the metadata that are present in the
+	// tree rooted in Op that the caller must drain. These can be
+	// colfetcher.ColBatchScan or colexec.Columnarizer operators that were
+	// created during NewColOperator call, but it also might include all of the
+	// sources from NewColOperatorArgs.MetadataSources if no metadata draining
+	// component was created.
+	MetadataSources execinfrapb.MetadataSources
 	// ToClose is a slice of components that need to be Closed.
 	ToClose     []colexecop.Closer
 	OpMonitors  []*mon.BytesMonitor

--- a/pkg/sql/colexec/materializer.go
+++ b/pkg/sql/colexec/materializer.go
@@ -72,7 +72,9 @@ type Materializer struct {
 // trailing metadata state, which is meant only for internal metadata
 // generation.
 type drainHelper struct {
-	sources      execinfrapb.MetadataSources
+	sources execinfrapb.MetadataSources
+	// If unset, the drainHelper wasn't Start()'ed, so all operations on it
+	// are noops.
 	ctx          context.Context
 	bufferedMeta []execinfrapb.ProducerMetadata
 	getStats     func() []*execinfrapb.ComponentStats
@@ -110,6 +112,10 @@ func (d *drainHelper) Start(ctx context.Context) {
 
 // Next implements the RowSource interface.
 func (d *drainHelper) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata) {
+	if d.ctx == nil {
+		// The drainHelper wasn't Start()'ed, so this operation is a noop.
+		return nil, nil
+	}
 	if d.bufferedMeta == nil {
 		d.bufferedMeta = d.sources.DrainMeta(d.ctx)
 		if d.bufferedMeta == nil {
@@ -127,6 +133,10 @@ func (d *drainHelper) Next() (rowenc.EncDatumRow, *execinfrapb.ProducerMetadata)
 
 // ConsumerDone implements the RowSource interface.
 func (d *drainHelper) ConsumerDone() {
+	if d.ctx == nil {
+		// The drainHelper wasn't Start()'ed, so this operation is a noop.
+		return
+	}
 	if d.getStats != nil {
 		// If getStats is non-nil, then the drainHelper is responsible for
 		// attaching the execution statistics to the span, yet we don't get the
@@ -255,12 +265,17 @@ func (m *Materializer) Child(nth int, verbose bool) execinfra.OpNode {
 // Start is part of the execinfra.RowSource interface.
 func (m *Materializer) Start(ctx context.Context) {
 	ctx = m.ProcessorBase.StartInternal(ctx, materializerProcName)
-	m.drainHelper.Start(ctx)
 	// We can encounter an expected error during Init (e.g. an operator
 	// attempts to allocate a batch, but the memory budget limit has been
 	// reached), so we need to wrap it with a catcher.
 	if err := colexecerror.CatchVectorizedRuntimeError(m.input.Init); err != nil {
 		m.MoveToDraining(err)
+	} else {
+		// Note that we intentionally only start the drain helper if
+		// initialization was successful - no starting the helper will tell it
+		// to not drain the metadata sources (which have not been properly
+		// initialized).
+		m.drainHelper.Start(ctx)
 	}
 }
 

--- a/pkg/sql/colflow/vectorized_flow.go
+++ b/pkg/sql/colflow/vectorized_flow.go
@@ -1081,24 +1081,29 @@ func (s *vectorizedFlowCreator) setupFlow(
 				return
 			}
 
-			// metadataSourcesQueue contains all the MetadataSources that need to be
+			// metadataSources contains all the MetadataSources that need to be
 			// drained. If in a given loop iteration no component that can drain
-			// metadata from these sources is found, the metadataSourcesQueue should be
-			// added as part of one of the last unconnected inputDAGs in
-			// streamIDToInputOp. This is to avoid cycles.
-			var metadataSourcesQueue []execinfrapb.MetadataSource
+			// metadata from these sources is found, the metadataSourcesQueue
+			// should be added as part of one of the last unconnected inputDAGs
+			// in streamIDToInputOp. This is to avoid cycles.
+			//
+			// The length of the slice is equal to the number of inputs (1 for
+			// most of processors, but could also be 0 or 2), and we keep the
+			// sources present in different input trees separately in order to
+			// prevent premature draining (in case of 2 inputs).
+			var metadataSources []execinfrapb.MetadataSources
 			// toClose is similar to metadataSourcesQueue with the difference that these
 			// components do not produce metadata and should be Closed even during
 			// non-graceful termination.
 			var toClose []colexecop.Closer
 			inputs := s.inputsScratch[:0]
 			for i := range pspec.Input {
-				input, metadataSources, closers, localErr := s.setupInput(ctx, flowCtx, pspec.Input[i], opt, factory)
+				input, ms, closers, localErr := s.setupInput(ctx, flowCtx, pspec.Input[i], opt, factory)
 				if localErr != nil {
 					err = localErr
 					return
 				}
-				metadataSourcesQueue = append(metadataSourcesQueue, metadataSources...)
+				metadataSources = append(metadataSources, ms)
 				toClose = append(toClose, closers...)
 				inputs = append(inputs, input)
 			}
@@ -1115,6 +1120,7 @@ func (s *vectorizedFlowCreator) setupFlow(
 				StreamingMemAccount:  s.newStreamingMemAccount(flowCtx),
 				ProcessorConstructor: rowexec.NewProcessor,
 				LocalProcessors:      localProcessors,
+				MetadataSources:      metadataSources,
 				DiskQueueCfg:         s.diskQueueCfg,
 				FDSemaphore:          s.fdSemaphore,
 				ExprHelper:           s.exprHelper,
@@ -1138,7 +1144,6 @@ func (s *vectorizedFlowCreator) setupFlow(
 			if flowCtx.EvalCtx.SessionData.TestingVectorizeInjectPanics {
 				result.Op = newPanicInjector(result.Op)
 			}
-			metadataSourcesQueue = append(metadataSourcesQueue, result.MetadataSources...)
 			if flowCtx.Cfg != nil && flowCtx.Cfg.TestingKnobs.CheckVectorizedFlowIsClosedCorrectly {
 				for _, closer := range result.ToClose {
 					func(c colexecop.Closer) {
@@ -1173,7 +1178,7 @@ func (s *vectorizedFlowCreator) setupFlow(
 			}
 
 			if err = s.setupOutput(
-				ctx, flowCtx, pspec, op, result.ColumnTypes, metadataSourcesQueue, toClose, factory,
+				ctx, flowCtx, pspec, op, result.ColumnTypes, result.MetadataSources, toClose, factory,
 			); err != nil {
 				return
 			}

